internal/utils/tfsdk.go

```diff
@@ -2,77 +2,165 @@ package utils
 
 import (
 	"context"
+	"encoding/json"
 
+	"github.com/hashicorp/terraform-plugin-framework-jsontypes/jsontypes"
 	"github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/path"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
+type ListMeta struct {
+	Index int
+	Path  path.Path
+	Diags diag.Diagnostics
+}
+
+type MapMeta struct {
+	Key   string
+	Path  path.Path
+	Diags diag.Diagnostics
+}
```

Comment:
The idea with using these, is to make the iter functions in each of these helpers a little less wordy. If you don't inline the function, you'll still have access to the parent metadata.